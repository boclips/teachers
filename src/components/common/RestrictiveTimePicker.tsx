import { Form } from '@ant-design/compatible';

import { Checkbox, TimePicker } from 'antd';
import { CheckboxChangeEvent } from 'antd/lib/checkbox';
import range from 'lodash/range';
import moment, { Moment } from 'moment';
import React from 'react';

interface Props {
  checkboxLabel: string;
  label: string;
  upperBound: number;
  onChange: (seconds: number) => void;
  initialValue: number;
}
interface State {
  value: Moment;
  disabled: boolean;
}

export class RestrictiveTimePicker extends React.Component<Props, State> {
  public state = {
    value: moment.utc(this.props.initialValue, 'X'),
    disabled: true,
  };

  private upperBound: {
    hours: number;
    seconds: number;
    minutes: number;
  };

  public constructor(props) {
    super(props);

    this.upperBound = this.boundToComponents(this.props.upperBound);
  }

  public componentDidUpdate(prevProps: Readonly<Props>): void {
    if (prevProps.upperBound !== this.props.upperBound) {
      this.upperBound = this.boundToComponents(this.props.upperBound);
    }
  }

  public render() {
    return (
      <section>
        <Form.Item label={this.props.checkboxLabel} className="hidden-label">
          <Checkbox onChange={this.handleCheckboxChange} />
        </Form.Item>
        <Form.Item label={this.props.label}>
          <TimePicker
            disabled={this.state.disabled}
            disabledHours={this.getDisabledHours}
            disabledMinutes={this.getDisabledMinutes}
            disabledSeconds={this.getDisabledSeconds}
            format={this.upperBound.hours > 0 ? 'hh:mm:ss' : 'mm:ss'}
            hideDisabledOptions={true}
            onChange={this.handleChange}
            onOpenChange={this.handleOpen}
            value={this.state.value}
          />
        </Form.Item>
      </section>
    );
  }

  private handleOpen = (open: boolean) =>
    setTimeout(() => {
      if (open) {
        // A bit hacky, but the only way we can reliably get the other options to show.
        // Antd won't give us a reference for the DOM element
        const selectContainers = document.querySelectorAll(
          '.ant-time-picker-panel-select',
        );
        if (selectContainers.length) {
          selectContainers.forEach((selectContainer) => {
            const selected = selectContainer.querySelector(
              '.ant-time-picker-panel-select-option-selected',
            ) as HTMLDivElement;

            let scrollTop = 0;
            if (selected) {
              scrollTop =
                selected.offsetTop +
                selected.clientHeight / 2 -
                selectContainer.clientHeight / 2;
            }

            selectContainer.scroll(0, Math.max(0, scrollTop));
          });
        }
      }
    }, 0);

  private boundToComponents = (bound: number) => ({
    hours: Math.floor(bound / 60 / 60),
    minutes: Math.floor((bound / 60) % 60),
    seconds: bound % 60,
  });

  private getDisabledHours = () =>
    range(24).filter((hour) => hour > this.upperBound.hours);

  private getDisabledMinutes = (selectedHour) =>
    range(60).filter((minute) => {
      if (selectedHour < this.upperBound.hours) {
        return false;
      }

      return minute > this.upperBound.minutes;
    });

  private getDisabledSeconds = (selectedHour, selectedMinute) =>
    range(60).filter((second) =>
      selectedMinute < this.upperBound.minutes ||
      selectedHour < this.upperBound.hours
        ? false
        : second > this.upperBound.seconds,
    );

  private handleChange = (time?: Moment) => {
    if (time == null) {
      return;
    }

    this.setState({ value: time }, () => {
      this.props.onChange(this.getSecondsFromValue());
    });
  };

  private getSecondsFromValue() {
    const time = this.state.value;

    // Moment is not immutable.
    const startOfTheDay = time.clone().startOf('d');
    const timeDifference = time.diff(startOfTheDay);

    return moment.duration(timeDifference).asSeconds();
  }

  private handleCheckboxChange = (event: CheckboxChangeEvent) => {
    this.setState({ disabled: !event.target.checked }, () => {
      if (this.state.disabled) {
        this.props.onChange(undefined);
      } else {
        this.props.onChange(this.getSecondsFromValue());
      }
    });
  };
}
