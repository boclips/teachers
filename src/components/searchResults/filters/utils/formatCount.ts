export function formatCount(count: number): string {
  if (count >= 500) {
    return `${count}+`;
  }
  return `${count}`;
}
