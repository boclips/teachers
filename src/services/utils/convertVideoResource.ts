import moment from 'moment';
import { StreamPlayback, Video, YoutubePlayback } from '../types/Video';

function getPlaybackProperties(
  resource: any,
): StreamPlayback | YoutubePlayback {
  if (resource.playback.streamUrl) {
    return new StreamPlayback(resource.playback.streamUrl);
  } else if (resource.playback.youtubeId) {
    return new YoutubePlayback(resource.playback.youtubeId);
  } else {
    throw Error(`No valid playback object found on resource: ${resource}`);
  }
}

export default function convertVideoResource(resource: any): Video {
  return {
    id: resource.id,
    title: resource.title,
    description: resource.description,
    duration: moment.duration(resource.playback.duration),
    releasedOn: new Date(resource.releasedOn),
    contentPartner: resource.contentPartner,
    thumbnailUrl: resource.playback.thumbnailUrl,
    playback: getPlaybackProperties(resource),
    subjects: resource.subjects,
  };
}
