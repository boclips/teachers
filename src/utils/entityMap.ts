export const organizeById = <T extends { id: string }>(
  collections: T[],
): { [key: string]: T } =>
  collections.reduce((result, collection) => {
    result[collection.id] = collection;
    return result;
  }, {});
