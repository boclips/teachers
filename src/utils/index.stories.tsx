import React from 'react';
import { Store } from 'redux';
import { Provider } from 'react-redux';
import { Router } from 'react-router';
import { createMemoryHistory } from 'history';
import BoclipsSecurity from 'boclips-js-security';
import State from '../types/State';
import { noOp } from './index';

export const storyWithProvider = (store: Store<State>) => (story) => (
  <Provider store={store}>{story()}</Provider>
);

export const storyWithRouter = () => (story) => (
  <Router history={createMemoryHistory()}>{story()}</Router>
);

export const storyWithAuthentication = () => (story) => {
  BoclipsSecurity.createInstance({
    clientId: 'testing-storybook',
    realm: 'testing-storybook',
    onLogin: noOp,
    onFailure: noOp,
    authEndpoint: '',
    mode: 'check-sso',
  });

  return story();
};
