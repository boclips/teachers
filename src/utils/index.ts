import classnames from 'classnames';

export const noOp: (args?: any) => any = () => {};

export const generateBorderRadiusClassNames = (
  currentIndex,
  columnCount,
  sizeOfArray,
) =>
  classnames({
    'border-radius--top-left': currentIndex === 0,
    'border-radius--top-right': currentIndex === columnCount - 1,
    'border-radius--bottom-left':
      currentIndex ===
      Math.ceil(sizeOfArray / columnCount) * columnCount - columnCount,
    'border-radius--bottom-right':
      currentIndex === Math.ceil(sizeOfArray / columnCount) * columnCount - 1,
  });
