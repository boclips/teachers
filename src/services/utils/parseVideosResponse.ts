import AppConfig from '../../app/AppConfig';
import { SearchResults } from '../../redux/State';
import convertVideoResource from './convertVideoResource';

export function parseVideosResponse(
  response: any,
  query: string,
): SearchResults {
  const videos = response.data._embedded.videos.map(convertVideoResource);
  const correlationId =
    response.headers[AppConfig.getCorrelationIdHeaderField()];

  return {
    videos,
    searchId: correlationId,
    query,
    paging: response.data.page,
  };
}
