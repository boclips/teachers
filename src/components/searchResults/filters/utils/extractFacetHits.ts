import { Facet } from 'src/types/VideoFacets';
import { MaxElementCount } from 'src/services/videos/parseVideosResponse';

export const extractFacetHits = (
  id: string,
  facet: { [id: string]: Facet },
): number => {
  if (!facet) {
    return 0;
  }

  if (!facet[id]) {
    return 0;
  }

  return facet[id].hits > MaxElementCount ? MaxElementCount : facet[id].hits;
};

export const extractTotalHits = (facets?: { [id: string]: Facet }): number =>
  facets
    ? Object.keys(facets)
        .map((key) => facets[key].hits)
        .reduce((acc, value) => acc + value, 0)
    : 0;
