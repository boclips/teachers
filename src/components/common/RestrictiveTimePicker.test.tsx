import { TimePicker } from 'antd';
import { shallow } from 'enzyme';
import difference from 'lodash/difference';
import range from 'lodash/range';
import * as moment from 'moment';
import * as React from 'react';
import { noOp } from '../../utils';
import { RestrictiveTimePicker } from './RestrictiveTimePicker';

it('matches the snapshot when rendering a restricted timepicker', () => {
  const component = shallow(
    <RestrictiveTimePicker
      checkboxLabel="Checkbox label"
      label="Label"
      upperBound={60}
      onChange={noOp}
      initialValue={0}
    />,
  );
  expect(component).toMatchSnapshot();
});

it('calls the onChange with seconds, when the time is changed', () => {
  const spy = jest.fn();

  const component = shallow(
    <RestrictiveTimePicker
      checkboxLabel="Checkbox label"
      label="Label"
      upperBound={60}
      onChange={spy}
      initialValue={0}
    />,
  );

  const timePicker = component.find(TimePicker);

  const onChange = timePicker.prop('onChange');

  onChange(moment.utc(63, 'X'), '');

  expect(spy).toHaveBeenCalledWith(63);
});

describe('upper bound restriction', () => {
  const testData = [
    {
      message: 'One hour',
      upperBound: 60 * 60,
      selectedHour: 0,
      selectedMinute: 0,
      expectedEnabledHours: [0, 1],
      expectedEnabledMinutes: range(60),
      expectedEnabledSeconds: range(60),
    },
    {
      message: 'One hour, selected one hour',
      upperBound: 60 * 60,
      selectedHour: 1,
      selectedMinute: 0,
      expectedEnabledHours: [0, 1],
      expectedEnabledMinutes: [0],
      expectedEnabledSeconds: [0],
    },
    {
      message: 'One hour and a half, selected one hour',
      upperBound: 60 * 60 * 1.5,
      selectedHour: 1,
      selectedMinute: 0,
      expectedEnabledHours: [0, 1],
      expectedEnabledMinutes: range(31),
      expectedEnabledSeconds: range(60),
    },
    {
      message: 'Two minutes 40',
      upperBound: 60 + 60 + 40,
      selectedHour: 0,
      selectedMinute: 0,
      expectedEnabledHours: [0],
      expectedEnabledMinutes: range(2 + 1),
      expectedEnabledSeconds: range(60),
    },
    {
      message: 'Two minutes 40, one minute selected',
      upperBound: 60 + 60 + 40,
      selectedHour: 0,
      selectedMinute: 1,
      expectedEnabledHours: [0],
      expectedEnabledMinutes: range(2 + 1),
      expectedEnabledSeconds: range(60),
    },
    {
      message: 'Two minutes 40, two minutes selected',
      upperBound: 60 + 60 + 40,
      selectedHour: 0,
      selectedMinute: 2,
      expectedEnabledHours: [0],
      expectedEnabledMinutes: range(2 + 1),
      expectedEnabledSeconds: range(40 + 1),
    },
  ];

  testData.forEach(data => {
    it(`restricts the input selection for ${data.message}`, () => {
      const component = shallow(
        <RestrictiveTimePicker
          checkboxLabel="Checkbox label"
          label="Label"
          upperBound={data.upperBound}
          onChange={noOp}
          initialValue={0}
        />,
      );

      const timePicker = component.find(TimePicker);

      const disabledHoursFn = timePicker.prop('disabledHours');

      expect(disabledHoursFn()).toEqual(
        difference(range(24), data.expectedEnabledHours),
      );

      const disabledMinutesFn = timePicker.prop('disabledMinutes');

      expect(disabledMinutesFn(data.selectedHour)).toEqual(
        difference(range(60), data.expectedEnabledMinutes),
      );

      const disabledSecondsFn = timePicker.prop('disabledSeconds');

      const disabledSeconds = disabledSecondsFn(
        data.selectedHour,
        data.selectedMinute,
      );
      expect(disabledSeconds).toEqual(
        difference(range(60), data.expectedEnabledSeconds),
      );
    });
  });
});
