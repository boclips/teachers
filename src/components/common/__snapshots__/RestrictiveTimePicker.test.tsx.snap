// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the snapshot when rendering a restricted timepicker 1`] = `
<section>
  <FormItem
    className="hidden-label"
    hasFeedback={false}
    label="Checkbox label"
  >
    <Checkbox
      indeterminate={false}
      onChange={[Function]}
    />
  </FormItem>
  <FormItem
    hasFeedback={false}
    label="Label"
  >
    <TimePicker
      align={
        Object {
          "offset": Array [
            0,
            -2,
          ],
        }
      }
      disabled={true}
      disabledHours={[Function]}
      disabledMinutes={[Function]}
      disabledSeconds={[Function]}
      focusOnOpen={true}
      format="mm:ss"
      hideDisabledOptions={true}
      onChange={[Function]}
      placement="bottomLeft"
      transitionName="slide-up"
      value={"1970-01-01T00:00:00.000Z"}
    />
  </FormItem>
</section>
`;
